### PR TITLE
Add explicit null checks for zero-offset struct fields

### LINKED ISSUE #77640: The Jit does not add explicit null checks for zero-offset field addresses
Reproduction:
```cs
Console.WriteLine((nint)Problem(null));

[MethodImpl(MethodImplOptions.NoInlining)]
static int* Problem(StructWithIndex* s)
{
    return &s->Index;
}

struct StructWithIndex
{
    public int Index;
    public int Value;
}
```
Compile and run.

Expected result: a `NullReferenceException` is thrown.

Actual result: `0` is printed.

Cause: odd code in `fgMorphField` that special-cases fields with zero offsets.

### PR DESCRIPTION
**Pull request by SingleAccretion**
Fixes #77640.

We expect some unavoidable regressions.

### CODE REVIEW

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -5171,55 +5171,21 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
...
-        // This means that we insert an explicit null check whenever we create byref by adding a
-        // constant offset to a ref, in a MACK_Addr context (meaning that the byref is not immediately
-        // dereferenced).  The alternative is "aggressive", which would not insert such checks (for
-        // small offsets); in this plan, we would transfer some null-checking responsibility to
-        // callee's of methods taking byref parameters.  They would have to add explicit null checks
-        // when creating derived byrefs from argument byrefs by adding constants to argument byrefs, in
-        // contexts where the resulting derived byref is not immediately dereferenced (or if the offset is too
-        // large).  To make the "aggressive" scheme work, however, we'd also have to add explicit derived-from-null
-        // checks for byref parameters to "external" methods implemented in C++, and in P/Invoke stubs.
-        // This is left here to point out how to implement it.
```
Comment (SingleAccretion) [inline src/coreclr/jit/morph.cpp:5184]: This comment is not correct; `ldflda` is required by the specification to throw on `null` addresses, immediately.

**Comment by SingleAccretion** [comment]
The [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=67299&view=ms.vss-build-web.run-extensions-tab) are relatively minor, the only large regressions are in MinOpts context from `coreclr_tests` (on some targets).

This fix will significantly simplify the `ADDR(FIELD)` => `FIELD_ADDR` transition.

@dotnet/jit-contrib

**Comment by EgorBo** [comment]
Reminds me that we used to tolerate this sort of issues and treat pointers to value types as never null.

**Comment by EgorBo** [comment]
Although the diffs don't look too bad in FullOpts so presumably it's fine?

**Review by EgorBo — APPROVED** [review]

**Comment by jakobbotsch** [comment]
Some of the diffs do look to be in some scary functions:
```scala
           3 (20.00 % of base) : 3668.dasm - ConfiguredValueTaskAwaiter[System.__Canon]:get_IsCompleted():bool:this
           3 (20.00 % of base) : 3851.dasm - ConfiguredValueTaskAwaiter[System.__Canon]:GetResult():System.__Canon:this
           3 (20.00 % of base) : 26284.dasm - System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder:SetException(System.Exception):this
           3 (20.00 % of base) : 16608.dasm - System.Runtime.CompilerServices.ValueTaskAwaiter`1[System.__Canon]:get_IsCompleted():bool:this
           3 (20.00 % of base) : 16609.dasm - System.Runtime.CompilerServices.ValueTaskAwaiter`1[System.__Canon]:GetResult():System.__Canon:this
           3 (10.00 % of base) : 16777.dasm - System.Runtime.CompilerServices.AsyncTaskMethodBuilder:AwaitUnsafeOnCompleted[YieldAwaiter,<MultipleSerial>d__23[System.__Canon]](byref,byref):this
           3 (0.53 % of base) : 14294.dasm - System.Collections.Generic.Dictionary`2[uint,System.__Canon]:FindValue(uint):byref:this
```

We should keep an eye on related benchmarks and/or double check that the diffs in these are not impactful.

**Comment by EgorBo** [comment]
> Some of the diffs do look to be in some scary functions:
> 
> ```scala
>            3 (20.00 % of base) : 3668.dasm - ConfiguredValueTaskAwaiter[System.__Canon]:get_IsCompleted():bool:this
>            3 (20.00 % of base) : 3851.dasm - ConfiguredValueTaskAwaiter[System.__Canon]:GetResult():System.__Canon:this
>            3 (20.00 % of base) : 26284.dasm - System.Runtime.CompilerServices.AsyncValueTaskMethodBuilder:SetException(System.Exception):this
>            3 (20.00 % of base) : 16608.dasm - System.Runtime.CompilerServices.ValueTaskAwaiter`1[System.__Canon]:get_IsCompleted():bool:this
>            3 (20.00 % of base) : 16609.dasm - System.Runtime.CompilerServices.ValueTaskAwaiter`1[System.__Canon]:GetResult():System.__Canon:this
>            3 (10.00 % of base) : 16777.dasm - System.Runtime.CompilerServices.AsyncTaskMethodBuilder:AwaitUnsafeOnCompleted[YieldAwaiter,<MultipleSerial>d__23[System.__Canon]](byref,byref):this
>            3 (0.53 % of base) : 14294.dasm - System.Collections.Generic.Dictionary`2[uint,System.__Canon]:FindValue(uint):byref:this
> ```
> 
> We should keep an eye on related benchmarks and/or double check that the diffs in these are not impactful.

I assume these all are small functions judging by the relative diffs and are inlined in the real world where null-checks might be folded

**Comment by SingleAccretion** [comment]
> I assume these all are small functions judging by the relative diffs and are inlined in the real world where null-checks might be folded

That would be my assumption as well. For reference, the large relative diffs mainly come from cases where have a method that immediately tail-calls another (a member function).

**Pull request merged**